src/Workspaces/Core/Portable/Execution/AssetBuilder.cs

```diff
@@ -75,6 +76,13 @@ public Asset Build(TextDocumentState state, SourceText text, CancellationToken c
             return asset;
         }
 
+        public Asset Build(OptionSet options, string language, CancellationToken cancellationToken)
+        {
+            // get around issue where this can't be captured in struct
+            var local = this;
```

Comment:
Yes.  

An alternate approach, that means you don't ever have to allocate, woudl be to allow GetOrCreateAsset to take an 'arg' parameter, this arg woudl then be passed to the lambda. 

This would allow the "this" and "language" to be captued in a ValueTuple that you could tehn pass to the lambda.  The lambda would then not capture any local state and could be cached. That would keep things alloc-free (which i presume is why you went with a struct in the first place).